src/goto-programs/remove_exceptions.cpp

```diff
@@ -98,175 +98,114 @@ class remove_exceptionst
+
+  bool may_throw_or_receive_exceptions(const goto_programt &) const;
 
   void instrument_exception_handler(
-    const irep_idt &function_id,
     goto_programt &goto_program,
-    const goto_programt::targett &);
+    const goto_programt::targett &,
+    bool may_catch);
 
   void add_exception_dispatch_sequence(
-    const irep_idt &function_id,
     goto_programt &goto_program,
     const goto_programt::targett &instr_it,
     const stack_catcht &stack_catch,
     const std::vector<exprt> &locals);
 
   void instrument_throw(
-    const irep_idt &function_id,
     goto_programt &goto_program,
     const goto_programt::targett &,
     const stack_catcht &,
     const std::vector<exprt> &);
 
   void instrument_function_call(
-    const irep_idt &function_id,
     goto_programt &goto_program,
     const goto_programt::targett &,
     const stack_catcht &,
     const std::vector<exprt> &);
 
   void instrument_exceptions(
-    const irep_idt &function_id,
     goto_programt &goto_program);
 };
 
-/// adds exceptional return variables for every function that may escape
-/// exceptions
-void remove_exceptionst::add_exceptional_returns(
-  const irep_idt &function_id,
-  goto_programt &goto_program)
+/// Create a global named java::\@inflight_exception that holds any exception
+/// that has been thrown but not yet caught.
+symbol_exprt remove_exceptionst::get_inflight_exception_global()
 {
+  const symbolt *existing_symbol =
+    symbol_table.lookup(INFLIGHT_EXCEPTION_VARIABLE_NAME);
+  INVARIANT(
+    existing_symbol != nullptr,
+    "Java frontend should have created @inflight_exception variable");
+  return existing_symbol->symbol_expr();
+}
 
-  auto maybe_symbol=symbol_table.lookup(function_id);
-  INVARIANT(maybe_symbol, "functions should be recorded in the symbol table");
-  const symbolt &function_symbol=*maybe_symbol;
-
-  // for now only add exceptional returns for Java
-  if(function_symbol.mode!=ID_java)
-    return;
-
-  if(goto_program.empty())
-    return;
-
-  // some methods (e.g. the entry method) have already been added to
-  // the symbol table; if you find it, initialise it
-  maybe_symbol=symbol_table.lookup(id2string(function_id)+EXC_SUFFIX);
-  if(maybe_symbol)
-  {
-    const symbolt &symbol=*maybe_symbol;
-    symbol_exprt lhs_expr_null=symbol.symbol_expr();
-    null_pointer_exprt rhs_expr_null((pointer_type(empty_typet())));
-    goto_programt::targett t_null=
-      goto_program.insert_before(goto_program.instructions.begin());
-    t_null->make_assignment();
-    t_null->source_location=
-      goto_program.instructions.begin()->source_location;
-    t_null->code=code_assignt(
-      lhs_expr_null,
-      rhs_expr_null);
-    t_null->function=function_id;
-    return;
-  }
-
-  // We generate an exceptional return value for any function that
-  // contains a throw or a function call that may escape exceptions.
+/// Checks whether a function may ever receive an exception, either by throwing
```

Comment:
I'm not clear what "receiving" an exception means? And if a exception can receive an exception by throwing, shouldn't this method just be called `may_receive_exceptions`?